### PR TITLE
feature(batch-action) Creating batch action finished notification

### PR DESCRIPTION
**Pull request by wiewiurdp**
# Description

Creating batch action finished notification.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Unit Tests
- [ ] Behat Test

# Checklist:
- [x] I have read the contribution requirements and fulfill them.
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: module/batch-action/src/Domain/Repository/BatchActionRepositoryInterface.php [hunk 1]
```diff
@@ -25,4 +25,6 @@ public function addEntry(BatchActionId $id, AggregateId $resourceId): void;
      * @param BatchActionMessage[] $messages
      */
     public function markEntry(BatchActionId $id, AggregateId $resourceId, array $messages): void;
+
+    public function isProcessEnded(BatchActionId $id): bool;
```
Comment (piotrkreft) [inline module/batch-action/src/Domain/Repository/BatchActionRepositoryInterface.php:29]:
```suggestion
    public function hasEnded(BatchActionId $id): bool;
```

File: module/batch-action/src/Infrastructure/Handler/BatchActionFinishedEventHandler.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        $user = $this->security->getUser();
+
+        if ($user) {
+            $userId = $user->getId();
+            $type = $this->translator->trans($event->getType()->getValue(), [], 'batch_action', 'en');
+            $notification = new BatchActionProcessFinishedNotification($type, $userId);
+            $notificationCommand = new SendNotificationCommand($notification, [$userId]);
+            $this->commandBus->dispatch($notificationCommand);
+        } else {
+            throw new AuthenticationException('User not set');
```
Comment (piotrkreft) [inline module/batch-action/src/Infrastructure/Handler/BatchActionFinishedEventHandler.php:48]: I'd ignore it I think - it's not that important.

File: module/batch-action/src/Infrastructure/Handler/EndBatchActionProcessCommandHandler.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    ) {
+        $this->repository = $repository;
+        $this->eventBus = $eventBus;
+        $this->commandBus = $commandBus;
+    }
+
+    public function __invoke(EndBatchActionProcessCommand $command): void
+    {
+        if ($this->repository->isProcessEnded($command->getId())) {
+            $event = new BatchActionFinishedEvent($command->getId(), $command->getType());
```
Comment (piotrkreft) [inline module/batch-action/src/Infrastructure/Handler/EndBatchActionProcessCommandHandler.php:39]: Please sync naming - once it's `ended`, other time `finished`

File: module/batch-action/src/Infrastructure/Handler/EndBatchActionProcessCommandHandler.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            $this->eventBus->dispatch($event);
+        } else {
+            $this->commandBus->dispatch($command, true);
```
Comment (piotrkreft) [inline module/batch-action/src/Infrastructure/Handler/EndBatchActionProcessCommandHandler.php:42]: As far as I remember we should throw an exception. @rprzedzik ?

File: module/batch-action/src/Infrastructure/Persistence/Repository/DbalBatchActionRepository.php [hunk 5]
```diff
@@ -140,6 +140,20 @@ public function markEntry(BatchActionId $id, AggregateId $resourceId, array $mes
         );
     }
 
+    public function isProcessEnded(BatchActionId $id): bool
+    {
+        $query = $this->connection->createQueryBuilder();
+
+        $result = $query->select("bool_and(success)")
```
Comment (piotrkreft) [inline module/batch-action/src/Infrastructure/Persistence/Repository/DbalBatchActionRepository.php:147]: it should check if not null, not the success itself  - false on success also means entry is processed

**Review by piotrkreft — CHANGES_REQUESTED** [review]

File: module/batch-action/src/Domain/Command/EndBatchActionProcessCommand.php [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+use Ergonode\BatchAction\Domain\ValueObject\BatchActionType;
+use Ergonode\SharedKernel\Domain\DomainCommandInterface;
+
+class EndBatchActionProcessCommand implements DomainCommandInterface
+{
+    private BatchActionId $id;
+
+    private BatchActionType $type;
+
+    public function __construct(BatchActionId $id, BatchActionType $type)
```
Comment (rprzedzik) [inline module/batch-action/src/Domain/Command/EndBatchActionProcessCommand.php:22]: BatchActionType is redundant you can fetch this information in handler by id.

In addition, if you have both values, there is a possibility that someone will use this command with an id of a different type than the type specified in the command, which will be an error

File: module/batch-action/src/Domain/Event/BatchActionEndedEvent.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+use Ergonode\BatchAction\Domain\ValueObject\BatchActionType;
+use Ergonode\SharedKernel\Domain\DomainEventInterface;
+
+class BatchActionEndedEvent implements DomainEventInterface
+{
+    private BatchActionId $id;
+
+    private BatchActionType $type;
+
+    public function __construct(BatchActionId $id, BatchActionType $type)
```
Comment (rprzedzik) [inline module/batch-action/src/Domain/Event/BatchActionEndedEvent.php:22]: Same like in command type property is redundant

File: module/batch-action/src/Infrastructure/Handler/BatchActionEndedEventHandler.php [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        $this->commandBus = $commandBus;
+    }
+
+    public function __invoke(BatchActionEndedEvent $event): void
+    {
+        $user = $this->security->getUser();
+
+        if ($user) {
+            $userId = $user->getId();
+            $type = $this->translator->trans($event->getType()->getValue(), [], 'batch_action', 'en');
```
Comment (rprzedzik) [inline module/batch-action/src/Infrastructure/Handler/BatchActionEndedEventHandler.php:43]: You can't translate messes before sending, especially using hardcoded languag!
You don't know what is user language ...
This should be done on when user fetch message to display.

File: module/batch-action/src/Domain/Notification/BatchActionProcessEndedNotification.php [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+declare(strict_types=1);
+
+namespace Ergonode\BatchAction\Domain\Notification;
+
+use Ergonode\Notification\Domain\NotificationInterface;
+use Ergonode\SharedKernel\Domain\Aggregate\UserId;
+
+class BatchActionProcessEndedNotification implements NotificationInterface
+{
+    private const MESSAGE = 'Batch action "%type%" ended';
```
Comment (rprzedzik) [inline module/batch-action/src/Domain/Notification/BatchActionProcessEndedNotification.php:17]: Add translations for this message

**Review by rprzedzik — CHANGES_REQUESTED** [review]

File: module/batch-action/src/Infrastructure/Handler/CreateBatchActionCommandHandler.php [hunk 10]
```diff
@@ -47,5 +48,7 @@ public function __invoke(CreateBatchActionCommand $command): void
             $entryCommand = new ProcessBatchActionEntryCommand($id, $type, $resourceId);
             $this->commandBus->dispatch($entryCommand, true);
         }
+        $endCommand = new EndBatchActionProcessCommand($id, $type);
```
Comment (wfajczyk) [inline module/batch-action/src/Infrastructure/Handler/CreateBatchActionCommandHandler.php:51]: Why You send `command` and next handle this command and send `event` ? 
I thing in this place You may send event  `BatchActionEndedEvent`

**Review by wfajczyk — CHANGES_REQUESTED** [review]

File: module/batch-action/src/Infrastructure/Handler/BatchActionEndedEventHandler.php [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        CommandBusInterface $commandBus
+    ) {
+        $this->translator = $translator;
+        $this->security = $security;
+        $this->commandBus = $commandBus;
+    }
+
+    public function __invoke(BatchActionEndedEvent $event): void
+    {
+        $user = $this->security->getUser();
```
Comment (wfajczyk) [inline module/batch-action/src/Infrastructure/Handler/BatchActionEndedEventHandler.php:39]: Will it work in asyncmode?

**Review by wfajczyk — CHANGES_REQUESTED** [review]

File: module/notification/src/Resources/translations/notification.en.yml [hunk 12]
```diff
@@ -0,0 +1,2 @@
+Batch action "%type%" ended: Batch action "%type%" ended
+product_delete: delete products
```
Comment (rprzedzik) [inline module/notification/src/Resources/translations/notification.en.yml:2]: This translation should be placed in product deletion batch action owner module

File: module/notification/src/Resources/translations/notification.pl.yml [hunk 13]
```diff
@@ -0,0 +1,2 @@
+Batch action "%type%" ended: Akcja masowa "%type%" zakończona
+product_delete: usuwanie produktów
```
Comment (rprzedzik) [inline module/notification/src/Resources/translations/notification.pl.yml:2]: This translation should be placed in product deletion batch action owner module

**Review by rprzedzik — CHANGES_REQUESTED** [review]

File: module/batch-action/src/Infrastructure/Handler/BatchActionEndedEventHandler.php (hunk 11 above)
Comment (wiewiurdp) [reply module/batch-action/src/Infrastructure/Handler/BatchActionEndedEventHandler.php:44]: after adding authorzaiton middleware (which I did in my last PR) it should work

File: module/batch-action/src/Infrastructure/Handler/BatchActionEndedEventHandler.php [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+        $batchAction = $this->batchActionRepository->load($event->getId());
+
+        if ($user && $batchAction) {
+            $userId = $user->getId();
+            $type = $this->translator->trans($batchAction->getType()->getValue(), [], 'notification', 'en');
```
Comment (wfajczyk) [inline module/batch-action/src/Infrastructure/Handler/BatchActionEndedEventHandler.php:49]: Translation only to `en` ?
I think is beter use 
```suggestion
            $type = $this->translator->trans($batchAction->getType()->getValue(), [], 'notification', '$user->getLanguage()`');
```

File: module/batch-action/src/Infrastructure/Handler/CreateBatchActionCommandHandler.php [hunk 15]
```diff
@@ -47,5 +48,7 @@ public function __invoke(CreateBatchActionCommand $command): void
             $entryCommand = new ProcessBatchActionEntryCommand($id, $type, $resourceId);
             $this->commandBus->dispatch($entryCommand, true);
         }
+        $endCommand = new EndBatchActionCommand($id);
+        $this->commandBus->dispatch($endCommand, true);
```
Comment (wfajczyk) [inline module/batch-action/src/Infrastructure/Handler/CreateBatchActionCommandHandler.php:52]: Why You send `command` and next handle this command and send `event` ?
I thing in this place You may send event `BatchActionEndedEvent`

**Review by wfajczyk — CHANGES_REQUESTED** [review]

File: module/product/src/Resources/translations/notification.en.yml [hunk 16]
```diff
@@ -0,0 +1,2 @@
+Batch action "%type%" ended: Batch action "%type%" ended
```
Comment (rprzedzik) [inline module/product/src/Resources/translations/notification.en.yml:1]: First translation should be in batch action module becous this is universal functionality 
Second line is connected to product delete batch action and that schold be in product module

File: module/product/src/Resources/translations/notification.pl.yml [hunk 17]
```diff
@@ -0,0 +1,2 @@
+Batch action "%type%" ended: Akcja masowa "%type%" zakończona
+product_delete: usuwanie produktów
```
Comment (rprzedzik) [inline module/product/src/Resources/translations/notification.pl.yml:2]: Firs translation should be in batch action module becous this is universal functionality 
Second line is connected to product delete batch action and that schold be in product module

**Review by rprzedzik — CHANGES_REQUESTED** [review]

File: module/batch-action/tests/Infrastructure/Handler/EndBatchActionProcessCommandHandlerTest.php [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+declare(strict_types=1);
+
+namespace Ergonode\BatchAction\Tests\Infrastructure\Handler;
+
+use Ergonode\BatchAction\Domain\Command\EndBatchActionCommand;
+use Ergonode\BatchAction\Infrastructure\Handler\EndBatchActionCommandHandler;
+use Ergonode\SharedKernel\Domain\Bus\EventBusInterface;
+use PHPUnit\Framework\TestCase;
+
+class EndBatchActionProcessCommandHandlerTest extends TestCase
```
Comment (piotrkreft) [inline module/batch-action/tests/Infrastructure/Handler/EndBatchActionProcessCommandHandlerTest.php:17]:
```suggestion
class EndBatchActionCommandHandlerTest extends TestCase
```

File: module/batch-action/src/Domain/Notification/BatchActionProcessEndedNotification.php [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Ergonode\BatchAction\Domain\Notification;
+
+use Ergonode\Notification\Domain\NotificationInterface;
+use Ergonode\SharedKernel\Domain\Aggregate\UserId;
+
+class BatchActionProcessEndedNotification implements NotificationInterface
```
Comment (piotrkreft) [inline module/batch-action/src/Domain/Notification/BatchActionProcessEndedNotification.php:15]:
```suggestion
class BatchActionEndedNotification implements NotificationInterface
```

File: module/batch-action/tests/Domain/Notification/BatchActionProcessEndedNotificationTest.php [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+
+declare(strict_types=1);
+
+namespace Ergonode\BatchAction\Tests\Domain\Notification;
+
+use Ergonode\BatchAction\Domain\Notification\BatchActionProcessEndedNotification;
+use Ergonode\SharedKernel\Domain\Aggregate\UserId;
+use PHPUnit\Framework\TestCase;
+
+class BatchActionProcessEndedNotificationTest extends TestCase
```
Comment (piotrkreft) [inline module/batch-action/tests/Domain/Notification/BatchActionProcessEndedNotificationTest.php:16]:
```suggestion
class BatchActionEndedNotificationTest extends TestCase
```

File: module/batch-action/tests/Infrastructure/Handler/EndBatchActionProcessCommandHandlerTest.php [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+
+    private EndBatchActionCommand $command;
+
+    protected function setUp(): void
+    {
+        $this->eventBus = $this->createMock(EventBusInterface::class);
+        $this->command = $this->createMock(EndBatchActionCommand::class);
+    }
+
+    public function testCommandHandlingProcessEnded(): void
```
Comment (piotrkreft) [inline module/batch-action/tests/Infrastructure/Handler/EndBatchActionProcessCommandHandlerTest.php:29]:
```suggestion
    public function testCommandHandlingBatchActionEnded(): void
```

**Review by piotrkreft — CHANGES_REQUESTED** [review]

File: module/batch-action/src/Infrastructure/Handler/CreateBatchActionCommandHandler.php (hunk 15 above)
Comment (wiewiurdp) [reply module/batch-action/src/Infrastructure/Handler/CreateBatchActionCommandHandler.php:52]: Command is saying that something is ended. This is not its responsibility what is going to be next

**Review by rprzedzik — APPROVED** [review]

**Review by piotrkreft — APPROVED** [review]

**Review by wfajczyk — APPROVED** [review]

**Pull request merged**